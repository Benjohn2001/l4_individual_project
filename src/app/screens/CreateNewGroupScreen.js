import React, { useEffect, useState } from 'react';
import { Text, View, Image, TextInput, SafeAreaView, Alert } from 'react-native';
import PressableIcon from '../components/PressableIcon';
import { COLOURS } from '../assets/colours';
import AppButtonPurple from '../components/AppButtonPurple';
import { getDatabase, set, ref, push } from 'firebase/database';
import { auth } from '../../firebase';
import randomColor from "randomcolor";


const CreateNewGroupScreen = ({ navigation }) => {
    const [groupName, setGroupName] = useState("")

    const createGroup = async () => {

        if(groupName==""){
            Alert.alert("Incomplete Form","One or more fields are empty")
        }else{
            const groupsRef = push(ref(getDatabase(), "/groups/"+auth.currentUser.uid))
            const groupMembersRef = push(ref(getDatabase(), "/groupMembers/"+groupsRef.key))
            const locationsRef = ref(getDatabase(), "/locations/"+groupsRef.key)
            const faceRef = ref(getDatabase(), "/clockFace/"+groupsRef.key)

            await set(groupsRef, {
                name: groupName,
                membersRef: groupsRef.key
            })
            await set(groupMembersRef,{
                member: auth.currentUser.uid,
                colour: randomColor()
            })
            await set(locationsRef,{
                locations: ["Uni","Work","Out","Free","Busy","Home"]
            })
            await set(faceRef,{
                background: "#FFFFFF"
            })
            setGroupName("")
            navigation.push("HomeScreen")
        }

    }
    return (
        <SafeAreaView className='flex-1 bg-primaryPurple'>
            <View className='flex-row justify-between pt-14 mx-10'>
                <PressableIcon
                        onPress={() => {
                            navigation.navigate("HomeScreen")
                        }}
                        icon="arrow-left"
                        size={40}
                        color="black"
                />
                <Text className="font-bold mr-auto ml-auto text-3xl">
                    New Group
                </Text>
            </View>
            <View className='justify-center items-center'>
                <Image 
                    className="w-44 h-44" 
                    source = {require("../assets/clock.png")}
                />
            </View>
            <View className='items-center'>
                <TextInput 
                    className="bg-secondaryPurple my-4 w-80 h-12 rounded-md" 
                    placeholder="Group Name" 
                    underlineColorAndroid = "transparent"
                    cursorColor={COLOURS.darkerPurple}
                    value={groupName}
                    onChangeText={(val) => setGroupName(val)}
                />
                <AppButtonPurple
                    title="Create"
                    onPress={() => {
                        createGroup()
                    }}
                />
            </View>
        </SafeAreaView>
    );
}
export default CreateNewGroupScreen;