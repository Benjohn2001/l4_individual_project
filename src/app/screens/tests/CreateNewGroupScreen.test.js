import renderer from "react-test-renderer";
import { render, screen, fireEvent } from '@testing-library/react-native';
import { Alert } from "react-native";
import CreateNewGroupScreen from "../CreateNewGroupScreen";

jest.mock('firebase/database', () => ({
    getDatabase: jest.fn().mockReturnThis(),
    push: jest.fn().mockReturnThis(),
    ref: jest.fn().mockReturnThis(),
    set: jest.fn().mockReturnThis(),
  }));

jest.mock('firebase/storage', () => ({
      getStorage: jest.fn().mockReturnThis(),
    }));

const navigation={
    navigate: jest.fn(),
    push: jest.fn()
}
  
jest.mock('firebase/app', () => ({
      auth:  jest.fn().mockReturnThis(),
      initializeApp: jest.fn().mockReturnThis(),
    }));

jest.mock('firebase/auth', () => ({
    auth:  jest.fn().mockReturnThis(),
      getAuth:  () => ({
          signOut:  () => ({
            then: () => ({
                catch: jest.fn()
            })
            
          }),
        }),
    }));



    

test('Check CreateNewGroupScreen working correctly',()=>{

    const {getByTestId } = render(
        <CreateNewGroupScreen navigation={navigation}/>
    );

    fireEvent.press(screen.getByTestId("pressableIcon"));
    expect(navigation.navigate).toBeCalledWith('HomeScreen')

    fireEvent.changeText(getByTestId("groupname"), "group")
    fireEvent.press(screen.getByText("Create"));
    // expect(navigation.push).toBeCalledWith('HomeScreen')

    const domTree = renderer.create(
        <CreateNewGroupScreen/>
        ).toJSON();
      expect(domTree).toMatchSnapshot();
})

jest.spyOn(Alert, 'alert');
test('Check CreateNewGroupScreen no groupname working correctly',()=>{

    render(
        <CreateNewGroupScreen navigation={navigation}/>
    );

    fireEvent.press(screen.getByTestId("pressableIcon"));
    expect(navigation.navigate).toBeCalledWith('HomeScreen')
    fireEvent.press(screen.getByText("Create"));
    expect(Alert.alert).toHaveBeenCalledTimes(1)
   
})



