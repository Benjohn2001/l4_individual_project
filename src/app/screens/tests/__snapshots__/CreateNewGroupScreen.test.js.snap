// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check CreateNewGroupScreen working correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "backgroundColor": "#C6C4FF",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Object {
          "justifyContent": "space-between",
        },
        Object {
          "paddingTop": 56,
        },
        Object {
          "marginLeft": 40,
          "marginRight": 40,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="pressableIcon"
      >
        <Text />
      </View>
    </View>
    <Text
      style={
        Array [
          Object {
            "fontWeight": "700",
          },
          Object {
            "marginRight": "auto",
          },
          Object {
            "marginLeft": "auto",
          },
          Object {
            "fontSize": 30,
            "lineHeight": 36,
          },
        ]
      }
    >
      New Group
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
        },
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "uri": "uri",
        }
      }
      style={
        Array [
          Object {
            "width": 176,
          },
          Object {
            "height": 176,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <TextInput
      cursorColor="#6B4EFF"
      onChangeText={[Function]}
      placeholder="Group Name"
      style={
        Array [
          Object {
            "backgroundColor": "#E7E7FF",
          },
          Object {
            "marginBottom": 16,
            "marginTop": 16,
          },
          Object {
            "width": 320,
          },
          Object {
            "height": 48,
          },
          Object {
            "borderBottomLeftRadius": 6,
            "borderBottomRightRadius": 6,
            "borderTopLeftRadius": 6,
            "borderTopRightRadius": 6,
          },
        ]
      }
      testID="groupname"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#6B4EFF",
          "borderBottomLeftRadius": 9999,
          "borderBottomRightRadius": 9999,
          "borderTopLeftRadius": 9999,
          "borderTopRightRadius": 9999,
          "height": 48,
          "marginLeft": 6,
          "marginRight": 6,
          "opacity": 1,
          "paddingTop": 12,
          "width": "83.333333%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
            },
            Object {
              "textAlign": "center",
            },
          ]
        }
      >
        Create
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
